tools/sql/conn.go

```diff
@@ -40,90 +39,99 @@ type (
 		driverName string
 	}
 	sqlConn struct {
+		driver   Driver
 		database string
 		db       *sqlx.DB
 	}
-)
 
-const (
-	defaultSQLPort              = 3306
-	readSchemaVersionSQL        = `SELECT curr_version from schema_version where db_name=?`
-	writeSchemaVersionSQL       = `REPLACE into schema_version(db_name, creation_time, curr_version, min_compatible_version) VALUES (?,?,?,?)`
-	writeSchemaUpdateHistorySQL = `INSERT into schema_update_history(year, month, update_time, old_version, new_version, manifest_md5, description) VALUES(?,?,?,?,?,?,?)`
-
-	createSchemaVersionTableSQL = `CREATE TABLE schema_version(db_name VARCHAR(255) not null PRIMARY KEY, ` +
-		`creation_time DATETIME(6), ` +
-		`curr_version VARCHAR(64), ` +
-		`min_compatible_version VARCHAR(64));`
-
-	createSchemaUpdateHistoryTableSQL = `CREATE TABLE schema_update_history(` +
-		`year int not null, ` +
-		`month int not null, ` +
-		`update_time DATETIME(6) not null, ` +
-		`description VARCHAR(255), ` +
-		`manifest_md5 VARCHAR(64), ` +
-		`new_version VARCHAR(64), ` +
-		`old_version VARCHAR(64), ` +
-		`PRIMARY KEY (year, month, update_time));`
+	// Driver is the driver interface that each SQL database needs to implement
+	Driver interface {
+		GetDriverName() string
+		CreateDBConnection(driverName, host string, port int, user string, passwd string, database string) (*sqlx.DB, error)
+		GetReadSchemaVersionSQL() string
+		GetWriteSchemaVersionSQL() string
+		GetWriteSchemaUpdateHistorySQL() string
+		GetCreateSchemaVersionTableSQL() string
+		GetCreateSchemaUpdateHistoryTableSQL() string
```

Comment:
[ NOT A BLOCKER ] Don't like the fact that there is an implicit assumption about the parameters of the SQL statement for the caller of this API. I think a ConnectionProvider interface and a ConnectionFactory as opposed to this driver interface would be more flexible & cleaner. However,  I understand it's unlikely that parameters would differ between SQL engines and this is something we can change later when we have a need for flexibility. So, up to you

Reply:
yeah, let's keep that in mind. It's also easier to manage that we give less flexibility at the beginning. It's also possible that people go to implement https://github.com/uber/cadence/blob/master/tools/common/schema/types.go#L52 this interface which will have most flexibility.